### PR TITLE
expression:Support collate for user session variable

### LINKED ISSUE #15168: UCP: Support collate for user session variable
## Description

```
mysql> set names utf8mb4 collate utf8mb4_general_ci;
Query OK, 0 rows affected (0.00 sec)

mysql> set @a = "a";
Query OK, 0 rows affected (0.00 sec)

mysql> select collation(@a);
+---------------+
| collation(@a) |
+---------------+
| utf8mb4_bin   |
+---------------+
1 row in set (0.00 sec)
```
The expected result is `utf8mb4_general_ci`.
A problem is that we use a map(string[string]) to store the session variable. If we try to support collation, we may need to change the map to `string[datum]`

## Score

* 50

## Mentor

* @wjhuang2016 

Contact the mentors: **#tidb-challenge-program** channel in [TiDB Community](https://join.slack.com/t/tidbcommunity/shared_invite/enQtNzc0MzI4ODExMDc4LWYwYmIzMjZkYzJiNDUxMmZlN2FiMGJkZjAyMzQ5NGU0NGY0NzI3NTYwMjAyNGQ1N2I2ZjAxNzc1OGUwYWM0NzE)  Slack Workspace

## Recommended Skills

* Golang programming

## Learning Materials

* English: [MySQL Document](https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_collation_connection)

### PR DESCRIPTION
**Pull request by SeaRise**
UCP #15168 
### What problem does this PR solve?
fix #15168 

### Check List 

Tests 

 - Integration test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change

### CODE REVIEW

**Comment by SeaRise** [comment]
@wjhuang2016 PTAL

**Comment by SeaRise** [comment]
@wjhuang2016 PTAL

**Comment by SeaRise** [comment]
@wjhuang2016 PTAL

File: expression/builtin_other.go [hunk 1]
```diff
@@ -710,10 +710,33 @@ func (c *getVarFunctionClass) getFunction(ctx sessionctx.Context, args []Express
...
 	bf := newBaseBuiltinFuncWithTp(ctx, args, types.ETString, types.ETString)
 	bf.tp.Flen = mysql.MaxFieldVarCharLength
+	if err := c.resolveCollation(ctx, args, &bf); err != nil {
+		return nil, err
+	}
 	sig = &builtinGetVarSig{bf}
 	return sig, nil
 }
 
+func (c *getVarFunctionClass) resolveCollation(ctx sessionctx.Context, args []Expression, bf *baseBuiltinFunc) (err error) {
```
Comment (wjhuang2016) [inline expression/builtin_other.go]: Why we need this function?

File: executor/set.go [hunk 2]
```diff
@@ -92,7 +92,12 @@ func (e *SetExecutor) Next(ctx context.Context, req *chunk.Chunk) error {
 				if err1 != nil {
 					return err1
 				}
-				sessionVars.Users[name] = fmt.Sprintf("%v", svalue)
+
+				if len(value.Collation()) != 0 {
+					sessionVars.Users[name] = types.NewCollationStringDatum(svalue, value.Collation(), collate.DefaultLen)
+				} else {
+					sessionVars.Users[name] = types.NewStringDatum(svalue)
```
Comment (wjhuang2016) [inline executor/set.go]: The collation  shouldn't be "", is this branch necessary?

File: expression/builtin_other.go (hunk 1 above)
Comment (SeaRise) [reply expression/builtin_other.go]: In `func (sf *ScalarFunction) Eval(row chunk.Row) (d types.Datum, err error)`
`d.SetValue(res, sf.RetType)`, so `d.Collate = sf.RetType`.

And `ScalarFunction.RetType = baseFunctionClass.getRetTp()` after `baseFunctionClass.getFunction` In `newFunctionImpl`.

File: executor/set.go (hunk 2 above)
Comment (SeaRise) [reply executor/set.go]: for sql `set @test_collate_var = 1`, collation will be "".
In mysql, `collation(test_collate_var )` return `binary`.
May be I should handle `TODO: we should consider the type of the argument, but not take it as string for all situations.` In `func (c *getVarFunctionClass) getFunction`

**Comment by SeaRise** [comment]
I'm very sorry, I'm going to give up this pr....

File: executor/set.go (hunk 2 above)
Comment (wjhuang2016) [reply executor/set.go]: We don't need to consider the type of the variable. Take it as string is enough.
What we need to do here is convert the string into a datum, the charset and collation can be got by `ctx.GetSessionVars().GetCharsetInfo()`
And your PR is almost complete,  please don't give up this pr.

Comment (SeaRise) [reply executor/set.go]: ok...Sorry for my  foolish operation.
I will open a new pr for it

Comment (wjhuang2016) [reply executor/set.go]: The implement of collation has many implications，feel free to ask me if you are not sure.

**Pull request closed without merging**